### DISCUSSION TITLE
Hiding Cursor on Desktop

### DISCUSSION THREAD

**Discussion by chadwyk** [axmolengine/axmol#2206] [Q&A]
What is the best way to hide the cursor? I'd like to create a custom cursor for Desktop apps by putting a sprite where the cursor location is. However I can't get the cursor to hide in Windowed mode or FullScreen mode.

I am using the following code on Mac:

```
    auto glView = ax::Director::getInstance()->getGLView();
    glView->setCursorVisible(false);
```

I do not know if this works or not on PC as I haven't tried yet. 

Is there a different way to do it?

**Comment by rh101** [comment]
The call to `glView->setCursorVisible(false);` does work on Windows.  You should trace that call on MacOS to see where it ends up.

To set a custom sprite as the cursor, and assuming MacOS uses GLFW too, then refer to the [GLFW documentation](https://www.glfw.org/docs/3.3/input_guide.html#cursor_custom).

For example (this does work on Windows):
```
auto glView = ax::Director::getInstance()->getGLView();
auto glViewImpl = dynamic_cast<GLViewImpl*>(glView);

auto* sourceImage  = new Image();
sourceImage->initWithImageFile("YourCursorImage.png"); // This *must* be in 32bit RGBA format.

GLFWimage glfwImage;
glfwImage.width  = sourceImage->getWidth();
glfwImage.height = sourceImage->getHeight();
glfwImage.pixels = sourceImage->getData();

GLFWcursor* cursor = glfwCreateCursor(&glfwImage, 0, 0);
glfwSetCursor(glViewImpl->getWindow(), cursor);
```

**Reply by chadwyk** [reply]
Hmm Yeah that code compiles but it doesn't change the cursor at all on Mac.  Then I tried changing it to a system cursor:

```
    auto glView = ax::Director::getInstance()->getGLView();
    auto glViewImpl = dynamic_cast<ax::GLViewImpl*>(glView);
    GLFWwindow* window = glViewImpl->getWindow();

    GLFWcursor* cursor = glfwCreateStandardCursor(GLFW_HAND_CURSOR);
        if (cursor) {
            glfwSetCursor(window, cursor);
            printf("System cursor set successfully.\n");
        } else {
            printf("Failed to create system cursor.\n");
        }
```

That didn't work either, but the logs say it was set.  I moved the function call to onMouseDown, and clicked it a whole bunch and it looked like I could see the cursor changing to the pointingHandCursor for maybe one frame, but then changes back.

I then wrote a function in a .mm file for metal that changes the cursor but that didn't work either.... Again I moved the function call to onMouseDown, and clicked it a whole bunch and it looked like I could see the cursor changing to the pointingHandCursor for maybe one frame, but then changes back. Again logging all says it should be working.

`        [[NSCursor pointingHandCursor] set];
`

Maybe something is overriding it and setting the cursor back to the default?

**Comment by chadwyk** [comment]
Hmmm some progress.... I think it is somehow related to my project.  

I created fresh new c++ and Lua apps and tried hiding the cursor and also changing to a different GLFW system cursor and both worked flawlessly. Weird.... Time to strip down my code and figure out what's preventing it! If I figure it out I'll comment it and close out the discussion.

**Comment by chadwyk** [comment]
Gah I figured it out! Apparently you can't change the cursor at all when using IMGui! ok... that shouldn't be a major issue for me as I am only using it for debugging now. But would be nice to find a work around.  I'll keep looking

**Comment by chadwyk** [comment]
Just doing a quick glance at ChatGPT, and it looks like IMGui has it's own cursor handling functions which you can set cursors or disable them or you can add your code to change it after it's rendering code. I'll play around with that. I'm marking this closed.
